pkg/cli/topo.go

```diff
@@ -0,0 +1,238 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"context"
+	"fmt"
+	"text/tabwriter"
+	"time"
+
+	"github.com/onosproject/onos-lib-go/pkg/cli"
+	"github.com/onosproject/onos-topo/api/topo"
+	"github.com/spf13/cobra"
+)
+
+func getGetEntityCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "get-entity <id>",
+		Args:  cobra.MinimumNArgs(1),
+		Short: "Get a topo entity",
+		RunE:  runGetEntityCommand,
+	}
+	/*
+		cmd.Flags().StringP("id", "i", "", "the id of the entity")
+		cmd.Flags().BoolP("verbose", "v", false, "whether to print the entity with verbose output")
+
+		_ = cmd.MarkFlagRequired("id")
+	*/
+
+	return cmd
+}
+
+func runGetEntityCommand(cmd *cobra.Command, args []string) error {
+
+	var objects []*topo.Object
+
+	outputWriter := cli.GetOutput()
+	writer := new(tabwriter.Writer)
+	writer.Init(outputWriter, 0, 0, 3, ' ', tabwriter.FilterHTML)
+
+	objects, err := readObjects(cmd, args)
+	if err != nil {
+		return err
+	}
+
+	if len(objects) != 0 {
+		switch obj := objects[0].Obj.(type) {
+		case *topo.Object_Entity:
+			_, _ = fmt.Fprintf(writer, "ID\t%s\n", objects[0].Ref.GetID())
+			_, _ = fmt.Fprintf(writer, "Type\t%s\n", obj.Entity.GetType())
+		case nil:
+			cli.Output("No object is set")
+			// No object is set
+		default:
+			cli.Output("get error")
+			// return ERROR
+		}
+	}
+	return writer.Flush()
+}
+
+func getAddEntityCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "add-entity <id> [args]",
+		Args:  cobra.MinimumNArgs(1),
+		Short: "Add an entity",
+		RunE:  runAddEntityCommand,
+	}
+	cmd.Flags().StringP("type", "t", "", "the type of the entity")
+	//_ = cmd.MarkFlagRequired("type")
+	return cmd
+}
+
+func runAddEntityCommand(cmd *cobra.Command, args []string) error {
+	return writeObject(cmd, args, topo.Object_ENTITY, topo.Update_INSERT)
+}
+
+func getGetRelationCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "get-relation <id>",
+		Args:  cobra.MinimumNArgs(1),
+		Short: "Get a topo relationship",
+		RunE:  runGetRelationCommand,
+	}
+
+	return cmd
+}
+
+func runGetRelationCommand(cmd *cobra.Command, args []string) error {
+
+	var objects []*topo.Object
+
+	outputWriter := cli.GetOutput()
+	writer := new(tabwriter.Writer)
+	writer.Init(outputWriter, 0, 0, 3, ' ', tabwriter.FilterHTML)
+
+	objects, err := readObjects(cmd, args)
+	if err != nil {
+		return err
+	}
+
+	if len(objects) != 0 {
+		switch obj := objects[0].Obj.(type) {
+		case *topo.Object_Relationship:
+			_, _ = fmt.Fprintf(writer, "ID\t%s\n", objects[0].Ref.GetID())
```

Comment:
go text templates can be better for outputting complex lines 
See example https://github.com/onosproject/onos-config/blob/92c9bb4d0b60e144350da7e09c264128bd046854/pkg/cli/models.go#L29

Reply:
thanks for the pointer. Will look into changing it. Dont understand how to use the text template when the outputting has a for loop:

  116         case *topo.Object_Relationship:
  117             _, _ = fmt.Fprintf(writer, "ID\t%s\n", objects[0].Ref.GetID())
  118             _, _ = fmt.Fprintf(writer, "type\t%s\n", obj.Relationship.GetType())
  119             for _, ref := range obj.Relationship.GetSourceRefs() {
  120                 _, _ = fmt.Fprintf(writer, "src-entity-id\t%s\n", string(ref.ID))
  121             }
  122             for _, ref := range obj.Relationship.GetTargetRefs() {
  123                 _, _ = fmt.Fprintf(writer, "tgt-entity-id\t%s\n", string(ref.ID))
  124             }

Reply:
@shadansari sorry for the late reply - the template can be created per line too, and/or be combined together. Like here
https://github.com/onosproject/onos-config/blob/92c9bb4d0b60e144350da7e09c264128bd046854/pkg/cli/snapshot.go#L29